### PR TITLE
Alter source-build job to use internal runtimes and nuget sources under a switch

### PR DESCRIPTION
**Pull request by mmitche**
- This enables us to use delegation SAS and ADO account tokens to access internal feeds instead of ambient secrets
- In addition, add a "enable-internal-runtimes" template using typical parameters. This will enable us to more easily make updates (e.g. storage account name) in the future.

### To double check:

* [ ] The right tests are in and the right validation has happened.  Guidance: https://github.com/dotnet/arcade/blob/main/Documentation/Validation.md

### CODE REVIEW

**Comment by mmitche** [comment]
Test job: https://dev.azure.com/dnceng/internal/_build/results?buildId=2463636&view=results

**Comment by mmitche** [comment]
Ignore the prebuilt failures.

**Comment by mmitche** [comment]
Updated

public run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=693147&view=results
internal run: https://dev.azure.com/dnceng/internal/_build/results?buildId=2463652&view=results

**Comment by mmitche** [comment]
One more round of tests, g2g I think:

internal: https://dev.azure.com/dnceng/internal/_build/results?buildId=2464208&view=results
public: https://dev.azure.com/dnceng-public/public/_build/results?buildId=694139&view=results

File: eng/common/core-templates/job/source-build.yml [hunk 1]
```diff
@@ -32,6 +32,11 @@ parameters:
   platform: {}
 
   is1ESPipeline: ''
+  # If set to true and running on a non-public project,
```
Comment (MichaelSimons) [inline eng/common/core-templates/job/source-build.yml:35]:
```suggestion

  # If set to true and running on a non-public project,
```

File: eng/common/core-templates/steps/enable-internal-runtimes.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+parameters:
+- name: federatedServiceConnection
+  type: string
+  default: 'dotnetbuilds-internal-read'
+- name: outputVariableName
+  type: string
+  default: 'dotnetbuilds-internal-container-read-token-base64'
+- name: expiryInHours
+  type: number
+  default: 1
```
Comment (MichaelSimons) [inline eng/common/core-templates/steps/enable-internal-runtimes.yml:13]: Where all are these creds used?  Are there concerns of this timing out?  CI legs are taking 1.5 hours regularly.

File: eng/common/templates-official/steps/enable-internal-runtimes.yml
Comment (MichaelSimons) [inline eng/common/templates-official/steps/enable-internal-runtimes.yml:1]: Why do we need a templates and templates-official version?  Where do either of these get used?

File: eng/common/core-templates/steps/enable-internal-runtimes.yml (hunk 2 above)
Comment (mmitche) [reply eng/common/core-templates/steps/enable-internal-runtimes.yml:13]: These are typically used when arcade installs the runtimes required for the repo build (and the referenced runtime is internal). In some isolated cases repos like installer would download runtime or aspnetcore zip files for redistribution. I think though, in source build, these end up always distributed via the source build intermediates so no issue there.

For source build, the internal runtime ends up installed after the inner build is done as part of tooling restore prior to creating source build intermediates. As long as the inner build doesn't last more than an hour this should be fine.

If we see a problem, we can up the timeout or come up with another solution (1 hour is the desired max) if tooling gets better, but we're pretty limited by available options right now. These tokens get passed to dotnet-install, which doesn't support anything aside from query string parameters. They also get used in testing etc. etc.

For full VMR builds, there should be no restore of internal bits.

File: eng/common/templates-official/steps/enable-internal-runtimes.yml
Comment (mmitche) [reply eng/common/templates-official/steps/enable-internal-runtimes.yml:1]: This is the format that main uses for templates post-1ESPT https://github.com/dotnet/arcade/blob/main/eng/common/template-guidance.md

A repo's official build should reference `eng/common/templates-official`, and public builds should reference `eng/common/templates`.

**Review by MichaelSimons — APPROVED** [review]

**Comment by mmitche** [comment]
Also, because the new flow is under a switch, we can migrate servicing branches (where breaks are more likely), incrementally.

**Pull request merged**